go-controller/pkg/factory/factory.go

```diff
@@ -499,30 +539,24 @@ func (wf *WatchFactory) addHandler(objType reflect.Type, namespace string, lsel
 		return true
 	}
 
-	// Process existing items as a set so the caller can clean up
-	// after a restart or whatever
-	existingItems := inf.inf.GetStore().List()
-	if processExisting != nil {
-		items := make([]interface{}, 0)
-		for _, obj := range existingItems {
-			if filterFunc(obj) {
-				items = append(items, obj)
-			}
+	inf.Lock()
+	defer inf.Unlock()
+
```

Comment:
@girishmg yes, that's the intention. This behavior is copied from the SharedInformer and prevents newly added handlers from missing events while they are being sent the initial ADD set. Consider an Update sent immediately after the GetStore().List(). Without RW locking the informer that Update could be delivered to existing handlers, and our newly added one would miss that event entirely and be left with an out-of-date version of the object. I think we had this race in the code before too.

I don't think this is a huge problem (but I'm happy to be corrected) because this only locks the same object type, most of high-scale types (pods, NetworkPolicy, etc) are filtered and take a lot less time, and the others (services, endpoints) only have one handler anyway.

Reply:
@dcbw your explanation makes sense to me.